lib/ansible/modules/system/parted.py

```diff
@@ -0,0 +1,654 @@
+
+
+def read_record(file_path, default=None):
+    """
+    Reads the first line of a file and returns it.
+    """
+    try:
+        f = open(file_path, 'r')
+        try:
+            return f.readline().strip()
+        finally:
+            f.close()
+    except IOError:
+        return default
+
+
+def get_unlabeled_device_info(device, unit):
+    """
+    Fetches device information directly from the kernel and it is used when
+    parted cannot work because of a missing label. It always returns a 'unknown'
+    label.
+    """
+    device_name = os.path.basename(device)
+    base = "/sys/block/{0}".format(device_name)
+
+    vendor      = read_record(base + "/device/vendor", "Unknown")
+    model       = read_record(base + "/device/model", "model")
+    logic_block = int(read_record(base + "/queue/logical_block_size", 0))
+    phys_block  = int(read_record(base + "/queue/physical_block_size", 0))
+    size_bytes  = int(read_record(base + "/size", 0)) * logic_block
+
+    size, unit  = format_disk_size(size_bytes, unit)
+
+    return {
+        'generic': {
+            'dev':            device,
+            'table':          "unknown",
+            'size':           size,
+            'unit':           unit,
+            'logical_block':  logic_block,
+            'physical_block': phys_block,
+            'model':          "{0} {1}".format(vendor, model),
+        },
+        'partitions': []
+    }
+
+
+def get_device_info(device, unit):
+    """
+    Fetches information about a disk and its partitions and it returns a
+    dictionary.
+    """
+    global module
+
+    # If parted complains about missing labels, it means there are no partitions.
+    # In this case only, use a custom function to fetch information and emulate
+    # parted formats for the unit.
+    label_needed = check_parted_label(device)
+    if label_needed:
+        return get_unlabeled_device_info(device, unit)
+
+    command = "parted -s -m {0} -- unit '{1}' print".format(device, unit)
+    rc, out, err = module.run_command(command)
+    if rc != 0 and 'unrecognised disk label' not in err:
+        module.fail_json(msg=(
+            "Error while getting device information with parted "
+            "script: '{0}'".format(command)),
+            rc=rc, out=out, err=err
+        )
+
+    return parse_partition_info(out, unit)
+
+
+def check_parted_label(device):
+    """
+    Determines if parted needs a label to complete its duties. Versions prior
+    to 3.1 don't return data when there is no label. For more information see:
+    http://upstream.rosalinux.ru/changelogs/libparted/3.1/changelog.html
+    """
+    # Check the version
+    parted_major, parted_minor = parted_version()
+    if (parted_major == 3 and parted_minor >= 1) or parted_major > 3:
+        return False
+
+    # Older parted versions return a message in the stdout and RC > 0.
+    rc, out, err = module.run_command("parted -s -m {0} print".format(device))
```

Comment:
Using format() here means it won't work with Python 2.6 and older. Since Ansible supports Python 2.4 and higher you should really be using the old formatting method.
See http://pyformat.info/